contracts/cw-proposal-single/src/contract.rs

```diff
@@ -711,9 +713,90 @@ pub fn query_info(deps: Deps) -> StdResult<Binary> {
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn migrate(_deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
-    // Don't do any state migrations.
-    Ok(Response::default())
+pub fn migrate(deps: DepsMut, env: Env, msg: MigrateMsg) -> Result<Response, ContractError> {
+    #[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+    struct BetaProposal {
+        pub title: String,
+        pub description: String,
+        /// The address that created this proposal.
+        pub proposer: Addr,
+        /// The block height at which this proposal was created. Voting
+        /// power queries should query for voting power at this block
+        /// height.
+        pub start_height: u64,
+        /// The minimum amount of time this proposal must remain open for
+        /// voting. The proposal may not pass unless this is expired or
+        /// None.
+        pub min_voting_period: Option<Expiration>,
+        /// The the time at which this proposal will expire and close for
+        /// additional votes.
+        pub expiration: Expiration,
+        /// The threshold at which this proposal will pass.
+        pub threshold: Threshold,
+        /// The total amount of voting power at the time of this
+        /// proposal's creation.
+        pub total_power: Uint128,
+        /// The messages that will be executed should this proposal pass.
+        pub msgs: Vec<CosmosMsg<Empty>>,
+
+        pub status: Status,
+        pub votes: Votes,
+        pub allow_revoting: bool,
+
+        /// Information about the deposit that was sent as part of this
+        /// proposal. None if no deposit.
+        pub deposit_info: Option<CheckedDepositInfo>,
+    }
+
+    match msg {
+        MigrateMsg::FromBeta {} => {
+            // Retrieve current map from storage
+            let current_map: Map<u64, BetaProposal> = Map::new("proposals");
+            let count = current_map
+                .keys(deps.storage, None, None, Order::Ascending)
+                .count();
+            println!("keys count {}", count);
+
+            let current = current_map
+                .range(deps.storage, None, None, Order::Ascending)
+                .collect::<StdResult<Vec<(u64, BetaProposal)>>>()?;
```

Comment:
To save gas, how would you feel about getting the `count` from this? Getting the length of a vector is constant time, whereas the way above is `O(N)` I think.

Reply:
hm but I do need to access the proposals in the old map so I can create the new structs -- unless you meant something else?

Reply:
I believe using .count() is linear here, we have encountered gas issues with it with howl

Reply:
the reason I had to collect here was because the closure below requires unique access to deps

Reply:
Sounds good! Thanks for the thoughts. :)